### PR TITLE
Default acc_dtype to floatX

### PR DESCRIPTION
**Pull request by tvwenger**
## Description

This PR changes the default data type for the internal accumulator to `config.floatX` when the input data type is `float`. The previous behavior always attempted to upcast `acc_dtype`, which caused the graph to include higher precision floats even when the user requested `floatX = 'float32'`.

## Related Issue

- [X] Closes https://github.com/pymc-devs/pytensor/issues/653

## Checklist

- [X] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [X] Added necessary documentation (docstrings and/or example notebooks)
- [X] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [X] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by tvwenger** [comment]
Some of the tests are failing: `TestRepeat` and `TestSubtensor`. I can see where the logic is failing based on the test error:

```

self = CAReduce(scalar_op=Composite{(i0 + Cast{float32}(i1))},axis=(0,),dtype=float32,acc_dtype=float32,upcast_discrete_output=True)
idtype = 'float64'

    def _acc_dtype(self, idtype):
        acc_dtype = self.acc_dtype
        if acc_dtype is None:
            return dict(
                bool="int64",
                int8="int64",
                int16="int64",
                int32="int64",
                uint8="uint64",
                uint16="uint64",
                uint32="uint64",
                float16=config.floatX,
                float32=config.floatX,
                complex64="complex128",
            ).get(idtype, idtype)
        elif acc_dtype in continuous_dtypes and idtype in discrete_dtypes:
            # Specifying a continuous accumulator for discrete input is OK
            return acc_dtype
        else:
            # The conversion has to be considered an upcast.
            upcasted_dtype = upcast(idtype, acc_dtype)
            if acc_dtype != upcasted_dtype:
>               raise TypeError(
                    f"Cannot build {self} node with input dtype {idtype} "
                    f"and acc_dtype {acc_dtype}, as precision would be lost. "
                    "To correct this error, you can:\n"
                    "  - not specify acc_dtype, or\n"
                    f"  - use an acc_dtype at least as precise as {upcasted_dtype}.\n"
                    '  - specify "dtype" instead of "acc_dtype", so '
                    "the reduction will be precise, but the result will "
                    'be casted into "dtype" at the end.\n'
                    "If you are expecting the precision loss, you can "
                    f'use tensor.cast(..., dtype="{acc_dtype}"), on your input.'
                )
E               TypeError: Cannot build CAReduce{Composite{(i0 + Cast{float32}(i1))}, axis=0} node with input dtype float64 and acc_dtype float32, as precision would be lost. To correct this error, you can:
E                 - not specify acc_dtype, or
E                 - use an acc_dtype at least as precise as float64.
E                 - specify "dtype" instead of "acc_dtype", so the reduction will be precise, but the result will be casted into "dtype" at the end.
E               If you are expecting the precision loss, you can use tensor.cast(..., dtype="float32"), on your input.
```

Although `CAReduce` is given the correct `dtype` and `acc_dtype`, the function `_acc_dtype` that is raising the exception isn't testing against the passed `dtype`, but rather the `dtype` of the input: `(i0 + Cast{float32}(i1))`. This must be resolving to `float64`.

**Comment by ricardoV94** [comment]
Yes I think the function can't and shouldn't downcast. If the input is float64 the same should be used for accumulation by default

**Comment by tvwenger** [comment]
@ricardoV94 I found out why `CAReduce` had a `float64` input despite `floatX="float32"`. The reason was that `verify_grad` in `pytensor/gradient.py` sets the random projection to `float64` no matter what, unless `cast_to_output_type=True`. I've updated the offending tests, so we'll see what happens.

I'd like to add a regression test using the MWE you provided. If you could point me to a good place to include such a test, I would be grateful!

**Comment by tvwenger** [comment]
There were more offending tests, so I changed the default in `verify_grad` instead.

File: tests/tensor/test_subtensor.py [hunk 1]
```diff
@@ -771,8 +771,16 @@ def test_grad_2d_inc_set_subtensor(self):
 
                 t = op(n[:z, :z], m)
                 gn, gm = pytensor.grad(pt_sum(t), [n, m])
-                utt.verify_grad(lambda m: op(n[:z, :z], m), [mv], mode=self.mode)
-                utt.verify_grad(lambda nn: op(nn[:z, :z], mv), [data], mode=self.mode)
+                utt.verify_grad(
+                    lambda m: op(n[:z, :z], m),
+                    [mv],
+                    mode=self.mode,
```
Comment (ricardoV94) [inline tests/tensor/test_subtensor.py:777]: If you remove the commas pre-commit will go back to the previous formatting and avoid these unrelated changes

```suggestion
                    mode=self.mode
```

File: tests/tensor/test_subtensor.py [hunk 2]
```diff
@@ -771,8 +771,16 @@ def test_grad_2d_inc_set_subtensor(self):
...
+                )
+                utt.verify_grad(
+                    lambda nn: op(nn[:z, :z], mv),
+                    [data],
+                    mode=self.mode,
```
Comment (ricardoV94) [inline tests/tensor/test_subtensor.py:782]: Same here

**Review by ricardoV94 — COMMENTED** [review]
Looks good just some nitpick

**Comment by ricardoV94** [comment]
For the regression test, probably in the same file where the rewrite is being tested

**Comment by tvwenger** [comment]
Well, we've reached an impasse I think. For some tests, unless we upcast the accumulator to `float64` as is currently implemented, then we lose precision and some of the gradient tests and precision tests fail. Consider the following test:

```python
     def test_mean_precision(self):
        # Check that the default accumulator precision is sufficient
        x = shared(np.asarray([1e8, 1, -1e8], dtype="float32"))
        m = x.mean()
        f = function([], m)
        m_val = f()
       assert np.allclose(m_val, 1.0 / 3)
```

The only way this test will succeed is if the internal accumulator uses `float64` despite the input and output types being `float32`.

**Comment by tvwenger** [comment]
Closed in favor of https://github.com/pymc-devs/pytensor/pull/656
